feat: implement proxy usage

BlackDark opened the pull request:
Closes #10 

Have written small changes for proxy usage.
Tested in my environment where i need a proxy.

Released necessary package and adapted changes to the github action:
- https://www.npmjs.com/package/@blackdark/hashicorp-js-releases
- https://github.com/BlackDark/setup-terraform/tree/proxy-custom-pkg

Tried using it in an action 

```yaml
      - name: Setup terraform
        uses: BlackDark/setup-terraform@proxy-custom-pkg
        with:
          terraform_version: ^1.0.0
```

Works:

![image](https://user-images.githubusercontent.com/9947112/129924099-4740bec0-a3a0-4490-b42b-9e7a4c63c659.png)

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/js-releases?pullRequest=11)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/js-releases?pullRequest=11)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)

<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/js-releases?pullRequest=11).</sub>

bahag-welscha commented:
What is the progress of this PR? I faced the same issue that the binaries can't be downloaded when a proxy is used.
With the Fork from @BlackDark i have no problem to download the binaries with active proxy.

duxbuse commented:
+1 please merge this

jpogran commented:
Hi @BlackDark, apologies for the delay in response. Thank you for the contribution! This looks like a good addition but there are a few things to get sorted first.

We had to update openpgp to latest stable in another PR, but that has made conflicting changes for you. Can you please rebase on main and fold those in?

It's hard to tell what changes are necessary to get proxy support added when you're reformatted the entire file. Can you please walk back the formatting changes so we can review this PR?

BlackDark commented:
> Hi @BlackDark, apologies for the delay in response. Thank you for the contribution! This looks like a good addition but there are a few things to get sorted first.
> 
> We had to update openpgp to latest stable in another PR, but that has made conflicting changes for you. Can you please rebase on main and fold those in?
> 
> It's hard to tell what changes are necessary to get proxy support added when you're reformatted the entire file. Can you please walk back the formatting changes so we can review this PR?

@jpogran Rebased and cleanup up commit

jpogran commented on src/utils.ts:
```diff
@@ -1,20 +1,34 @@
+let proxyConf = {};
+
+if (httpProxy || httpsProxy) {
+  console.log(`Found proxy setting. Using to download files.`);
+  console.log(`HTTP: ${httpProxy}`);
+  console.log(`HTTPS: ${httpProxy}`);
```
```suggestion
```

We shouldn't be printing to the console proxy details

BlackDark replied:
Removed

jpogran commented on src/utils.ts:
```diff
@@ -1,20 +1,34 @@
+  url: string,
+  options: AxiosRequestConfig = {}
+): Promise<any> {
+  console.log(`Request to: ${url} with options: ${JSON.stringify(options)}`);
```
```suggestion
```
We shouldn't be printing to the console proxy details

BlackDark replied:
What would be the best solution to add debug functionality to check if the correct proxy is used?
The default set up job (on Github Enterprise) also prints proxy configs if setup:
![image](https://user-images.githubusercontent.com/9947112/142840735-412740f5-e2be-4dae-9b91-82391eaed93a.png)

jpogran replied:
This is a good question, and is something that has some variance in implementations. In this case we expect the application using this library to output information like this, not the library itself.

Logging information is an important task of any library, but there isn't a way in this library to currently log anything. A future PR could introduce an interface and method overload that allows logging to different targets, but this is out of scope here.

BlackDark replied:
Removed

jpogran commented on src/utils.ts:
```diff
@@ -1,20 +1,34 @@
+    const result = await axios.get(url, { ...options });
+    return result.data;
+  } catch (e: any) {
+    console.error(e.toJSON());
+    throw new Error(`Request for terraform binary failed: ${e && e.message}`);
```
Going to ask internally about the usage for this library, but I think we should follow the previous pattern of the resolve/reject with promises instead of try/catch and printing a message to the console.

jpogran replied:
Reviewed with the team and we agree the approach here is good with the exception of catching the error here. We shouldn't print the error to the console and throw a new formatted here.

This library is used for more than downloading terraform, so instead we should let the caller handle the exception.

BlackDark replied:
Remove the try catch so the caller must handle the error

jpogran reviewed (changes requested):
Minor request not to print proxy details to the console. Going to look into how this library is handling errors versus the presentation here.

jpogran commented:
Looking good so far. If you could add a short note to the README about the environment variables needed in order to activate this functionality I'd appreciate it. If not I can do that as a follow up.

BlackDark commented:
> Looking good so far. If you could add a short note to the README about the environment variables needed in order to activate this functionality I'd appreciate it. If not I can do that as a follow up.

Hi @jpogran found some time working on my PR. Updated your comments as suggested and updated the README.

dbanck commented:
Thanks a lot for the updates.

I noticed one thing when running npm install.
```
found 8 vulnerabilities (6 moderate, 1 high, 1 critical)
  run `npm audit fix` to fix them, or `npm audit` for details
```
Most of the vulnerabilities are related to dev packages and not part of this PR, but the critical one is. It was introduced by a dependency of `proxy-agent` and can be fixed by running `npm update vm2 --depth 5`.

Do you mind updating the deps accordingly?

BlackDark commented:
> npm update vm2 --depth 5

Updated the dependency :)

jpogran reviewed (approved):
Thank you for contributing this!

jpogran commented:
Apologies on the delay, merging this now!

magnetikonline commented:
Hey @BlackDark / @jpogran - I've been tracking an annoying deprecation warning I've been seeing in the https://github.com/hashicorp/setup-terraform GitHub Action recently.

```
(node:1617) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
(Use `node --trace-deprecation ...` to show where the warning was created)
```

This is actually seen in the opening screenshot 😄.

Tracking this back, it's been this change that's cause the message:

- Inclusion of `proxy-agent` package.
- This in-turn uses `pac-proxy-agent`
- This in-turn uses `get-uri`.
- This in-turn uses `ftp` (phew!)
- The (very old) `ftp` package makes use of `new Buffer()`: https://github.com/mscdex/node-ftp/pull/230

Got there!

I wonder if there is an alternative to the `proxy-agent` package?

jpogran commented:
@magnetikonline I've created a new issue for request with the content you've posted. We'll look into seeing what can be done about this..

Next time, you can open a new issue and link to the PR in question instead of posting to a closed PR. Not all projects keep tabs on closed PRs.

magnetikonline commented:
> Next time, you can open a new issue and link to the PR in question instead of posting to a closed PR. Not all projects keep tabs on closed PRs.

Fair call @jpogran - sorry about that.
